### PR TITLE
Add tags functionality

### PR DESCRIPTION
**Pull request by thebigmunch**
See #107 for previous discussion.

### CODE REVIEW

**Comment by darrenburns** [comment]
This is _very_ exciting, I really like how it looks 😊 

> So, I finally came back to this and created a branch implementing cucumber tag expressions in this branch.
> 
> I actually thought of a syntax I use somewhere else to use the style in my tags branch without the (exclude-)tags options being multiple. But, I think the advantages of the cucumber tag expressions outweigh having to quote them:
> * They are maintained/documented by someone else (super plus).
    Caveat: it's documented as using @ to denote the tags, but that isn't actually supported. So, their documentation is wrong/out-of-date?
> * They support far more complex scenarios for those edge case users. 
> * They don't require multiple-style options.
> * Did I mention the parser doesn't need to be maintained/documented by any of us? ^_^

Regarding the "@" to denote tags, I'm assuming they just internally prepend `@` to the start of each tag? Within Ward itself, whenver a tag is encountered we could do this and then the tag expressions supplied via command line would presumably work with `@`? Can you think of any benefits of supporting the `@` before tag names? I feel like I prefer it but am struggling to understand _why_ I prefer it.

I'm a big fan of avoiding `multiple` options too.

**Comment by thebigmunch** [comment]
> Regarding the "@" to denote tags, I'm assuming they just internally prepend `@` to the start of each tag? Within Ward itself, whenver a tag is encountered we could do this and then the tag expressions supplied via command line would presumably work with `@`? Can you think of any benefits of supporting the `@` before tag names? I feel like I prefer it but am struggling to understand _why_ I prefer it.

From what I can see, the ``@`` was actually a v1 thing that they never updated the docs for when they went to v2. The ``cucumber-tag-expressions`` only supports v2.

**Comment by thebigmunch** [comment]
The best solution might just be to copy/paste the help from the Cucumber docs into whatever tagging docs go into ward while removing the @'s. Then remove the link to the Cucumber docs from the option help.

**Comment by darrenburns** [comment]
I probably won't copy/paste them, but would rewrite them to include Ward specific examples, and link back to the original source.

**Comment by thebigmunch** [comment]
> I probably won't copy/paste them, but would rewrite them to include Ward specific examples, and link back to the original source.

Well, the only non-ward thing in the [tag-expressions](https://cucumber.io/docs/cucumber/api/#tag-expressions) doc is the word ``Scenarios``, so the only thing necessary would be to change that to ``Tests``. Whether you wanted to use different tag names or not is up to you, but the tagging syntax is a generic thing. It just needs to be shown. And, of course, I didn't mean a strict copy/paste : P

Most of the rest of the tags docs would be showing the syntax examples in use with ward-style tests.

**Comment by thebigmunch** [comment]
PS I should probably mention that I've had my [audio-metadata](https://github.com/thebigmunch/audio-metadata) project running on my tags branch for a couple weeks now: https://github.com/thebigmunch/audio-metadata/commit/5ea3e7dfa6b1e2af76a4feadeb89e8471161a03e. More if you count locally.

**Comment by darrenburns** [comment]
Oh brilliant, how have you been finding working with it? Any thoughts or concerns on how tagging feels to use, how it integrates with Ward, or Ward in general?

**Comment by thebigmunch** [comment]
> Oh brilliant, how have you been finding working with it? Any thoughts or concerns on how tagging feels to use, how it integrates with Ward, or Ward in general?

Been loving both ward and the tagging functionality. Truthfully, it's stupid simple yet utterly flexible and powerful. So much better than plain search for the use cases I have.

The only issue I've had with the cucumber tag expressions is with passing the quoted expressions to nox as posargs. But, this isn't an issue on ward's end. Not even really an issue on nox's end. That's just what happens with the terminal input. It may or may not affect other things like tox.

**Comment by darrenburns** [comment]
Nice, glad to hear it. I really like this syntax and would love to push forward with this. If you need help at all let me know!

**Comment by darrenburns** [comment]
I've just merged in master so that this is up to date with the Poetry migration

I just played around with this and love it! So simple, easy to use, and obvious how it works even without docs.

File: ward/collect.py [hunk 1]
```diff
@@ -115,21 +126,39 @@ def get_tests_in_modules(
...
+                    tags=meta.tags or [],
                 )
 
 
 def search_generally(
-    tests: Iterable[Test], query: str = ""
+    tests: Iterable[Test],
+    query: str = "",
+    tags: Optional[Expression] = None,
+    exclude_tags: Optional[Expression] = None,
```
Comment (darrenburns) [inline ward/collect.py]: What's the use-case for `exclude_tags`? Can't you just use `not` in your query?

Comment (thebigmunch) [reply ward/collect.py]: Indeed. I'll remove that the next time I work on it.

**Comment by thebigmunch** [comment]
I rebased and force-pushed along with removing ``exclude_tags``.

Forgot to update the lock file on the first one ^_^

**Comment by thebigmunch** [comment]
PS If you or anyone else gets the urge to document and add tests for this, feel free. I'm not sure when I'll be able to get around to doing it myself.

**Comment by darrenburns** [comment]
No problem, I'm happy to add tests and documentation

**Comment by darrenburns** [comment]
I added some tests and will now add some documentation :)

**Pull request merged**